gossip/src/duplicate_shred_handler.rs

```diff
@@ -134,133 +96,95 @@ impl DuplicateShredHandler {
-
-    fn dump_pubkeys_with_low_stakes(
-        cached_staked_nodes: &HashMap<Pubkey, u64>,
-        slot_chunk_map: &mut SlotChunkMap,
-    ) {
-        let mut stakes_and_keys: Vec<(u64, Pubkey)> = slot_chunk_map
-            .keys()
-            .map(|k| (cached_staked_nodes.get(k).copied().unwrap_or_default(), *k))
-            .collect();
-        stakes_and_keys.select_nth_unstable(MAX_PUBKEY_PER_SLOT / 2);
-        for (_, key) in stakes_and_keys {
-            slot_chunk_map.remove(&key);
-        }
+        Ok(())
     }
 
-    fn mark_slot_proof_received(&mut self, slot: u64) {
-        self.chunk_map.insert(slot, SlotStatus::Frozen);
-        for (_, current_slots_set) in self.validator_pending_proof_map.iter_mut() {
-            current_slots_set.remove(&slot);
-        }
+    fn should_consume_slot(&self, slot: Slot) -> bool {
+        slot > self.last_root
+            && slot < self.last_root.saturating_add(self.cached_slots_in_epoch)
+            && !self.consumed.contains(&slot)
+            && !self.blockstore.has_duplicate_shreds_in_slot(slot)
     }
 
-    fn insert_chunk(&mut self, data: DuplicateShred) -> Option<Vec<DuplicateShred>> {
-        let chunk_map_entry = self
-            .chunk_map
-            .entry(data.slot)
-            .or_insert_with(|| SlotStatus::UnfinishedProof(HashMap::new()));
-        let slot_chunk_map = match chunk_map_entry {
-            SlotStatus::Frozen => return None,
-            SlotStatus::UnfinishedProof(slot_chunk_map) => slot_chunk_map,
-        };
-        let proof_chunk_map = slot_chunk_map
-            .entry(data.from)
-            .or_insert_with(|| ProofChunkMap::new(data.num_chunks()));
-        if data.num_chunks() != proof_chunk_map.num_chunks
-            || data.chunk_index() >= proof_chunk_map.num_chunks
-        {
-            return None;
+    fn maybe_prune_buffer(&mut self) {
+        if self.buffer.len() < BUFFER_CAPACITY.saturating_mul(2) {
+            return;
         }
-        let slot = data.slot;
-        let from = data.from;
-        match proof_chunk_map
-            .chunks
-            .get_mut(usize::from(data.chunk_index()))
-        {
-            None => return None,
-            Some(entry) if entry.is_some() => return None,
-            Some(entry) => *entry = Some(data),
+        self.consumed.retain(|slot| *slot > self.last_root);
+        // Filter out obsolete slots and limit number of entries per pubkey.
+        let buffer: Vec<((Slot, Pubkey), BufferEntry)> = {
+            let mut counts = HashMap::<Pubkey, usize>::new();
+            self.buffer
+                .drain()
+                .filter(|((slot, _), _)| *slot > self.last_root && !self.consumed.contains(slot))
+                .filter(|((_, pubkey), _)| {
+                    let count = counts.entry(*pubkey).or_default();
+                    *count = count.saturating_add(1);
+                    *count <= MAX_NUM_ENTRIES_PER_PUBKEY
+                })
+                .collect()
```

Comment:
Just curious, is there a performance difference if you only collect the keys into a vector and then filter keys in hashmap? Because hashmap deletion should be O(1). I was mostly concerned about size of BufferEntry, but I guess there is no copy here?

Reply:
Removed the vector for the 1st part of pruning.
For the 2nd part we need a vector for `select_nth_unstable` anyways.
I think the code would be just cleaner if we `drain` the hashmap and then `collect` back remaining entries into a hashmap at the end.
Performance-wise probably does not matter either way.